### PR TITLE
Add support for Gradle Configuration Cache

### LINKED ISSUE #7299: Incompatibility issue with Gradle configuration cache
The configuration cache is a feature that significantly improves build performance by caching the result of the configuration phase and reusing this for subsequent builds.

Implementing support for the configuration cache in this plugin would improve users experience by dramatically reducing build times.

See https://docs.gradle.org/current/userguide/configuration_cache.html

The configuration cache is already available since Gradle 6.6.

When I run `./gradlew app:assembleRelease`, and the configuration doesn't come from the cache, Gradle reports this error:

```
- Task `:app:transformClassesWithRealmTransformerForDebug` of type `com.android.build.gradle.internal.pipeline.TransformTask`: cannot serialize object of type 'org.gradle.api.internal.project.DefaultProject', a subtype of 'org.gradle.api.Project', as these are not supported with the configuration cache.
  See https://docs.gradle.org/6.8.1/userguide/configuration_cache.html#config_cache:requirements:disallowed_types
```

Are there any plans to support configuration cache?

### PR DESCRIPTION
**Pull request by cmelchior**
Closes #7299 

This PR adds support for the Gradle Configuration Cache by moving all access to the `Project` instance to evaluation time which allows it to be cached for execution. 

This mostly impacted how analytics are done, so I had to refactor this. As part of that, I moved most of the code to Kotlin.
I thought about splitting analytics out entirely to a Build Service, but it would mean that analytics don't run if people are not using the Gradle Plugin (which some do and we have guides on how to go about it).

So this PR is compatible with people both using the Gradle Plugin and configuring dependencies/transformer themselves.

### CODE REVIEW

File: realm-transformer/build.gradle [hunk 1]
```diff
@@ -94,3 +93,9 @@ java {
     withSourcesJar()
     withJavadocJar()
 }
+
+compileKotlin {
+    kotlinOptions {
+        freeCompilerArgs = ["-Xinline-classes"]
```
Comment (rorbech) [inline realm-transformer/build.gradle:99]: Should be stable from Kotlin 1.5, so could also just bump Kotlin 🙈

File: realm-transformer/src/main/kotlin/io/realm/analytics/RealmAnalytics.kt [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+        return true
+    }
+
+    private fun isAnalyticsEnabled(project: Project): Boolean {
+        val env = System.getenv()
+        return !project.gradle.startParameter.isOffline
+                && env["REALM_DISABLE_ANALYTICS"] == null
+                && env["CI"] == null
+    }
+}
```
Comment (rorbech) [inline realm-transformer/src/main/kotlin/io/realm/analytics/RealmAnalytics.kt:150]:
```suggestion
}

```

File: realm-transformer/src/main/kotlin/io/realm/analytics/ComputerIdentifierGenerator.kt [hunk 3]
```diff
@@ -0,0 +1,101 @@
...
+                        break
+                    }
+                }
+            } finally {
+                `is`.close()
+            }
+            return if (result == null) UNKNOWN else Utils.hexStringify(Utils.sha256Hash(result.toByteArray()))
+        }
+    }
+}
```
Comment (rorbech) [inline realm-transformer/src/main/kotlin/io/realm/analytics/ComputerIdentifierGenerator.kt:101]:
```suggestion
}

```

**Review by rorbech — APPROVED** [review]
LGTM. Remember to set `REALM_DISABLE_ANALYTICS=true` in the CI build.

**Comment by Jeff11** [comment]
@cmelchior since this was merged a year ago, I wonder why it's not yet in a release build.

This cr link gives a 404 error https://github.com/realm/realm-java/tree/cr/fix-java11-build-errors

**Pull request merged**